internals/api/encrypted_value.go

```diff
@@ -0,0 +1,93 @@
+package api
+
+import (
+	"encoding/json"
+
+	"github.com/secrethub/secrethub-go/internals/crypto"
+)
+
+var (
+	ErrInvalidEncryptionType = errAPI.Code("invalid_encryption_type").Error("invalid encryption method provided for key")
+)
+
+type EncryptionType string
+
+const (
+	EncryptionTypeRSAAES  EncryptionType = "rsa-oaep-aes-gcm"
+	EncryptionTypeAWSKKMS EncryptionType = "aws-kms"
+)
+
+type EncryptedValue struct {
+	EncryptionType *EncryptionType `json:"encryption_type"`
+	Ciphertext     []byte          `json:"ciphertext"`
+	Metadata       interface{}     `json:"metadata"`
+}
+
+type EncryptionMetadataRSAAES struct {
+	RSAKeySize      *int   `json:"rsa_key_size"`
+	AESKeySize      *int   `json:"aes_key_size"`
+	EncryptedAESKey []byte `json:"aes_key"`
+	AESNonce        []byte `json:"aes_nonce"`
+}
+
+type EncryptionMetadataAWSKMS struct {
+}
+
+func (ek EncryptedValue) UnmarshalJSON(b []byte) error {
+	encodedMetadata := json.RawMessage{}
+	ek.Metadata = &encodedMetadata
+	err := json.Unmarshal(b, &ek)
+	if err != nil {
+		return err
+	}
+
+	if ek.EncryptionType == nil {
+		return ErrInvalidEncryptionType
+	}
+
+	switch *ek.EncryptionType {
+	case EncryptionTypeRSAAES:
+		ek.Metadata = &EncryptionMetadataRSAAES{}
+	case EncryptionTypeAWSKKMS:
+		ek.Metadata = &EncryptionMetadataAWSKMS{}
+	default:
+		return ErrInvalidEncryptionType
+	}
+
+	err = json.Unmarshal(encodedMetadata, ek.Metadata)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ek EncryptedValue) ToCiphertextRSAAES() (*crypto.CiphertextRSAAES, error) {
```

Comment:
Ideally, I don't think this conversion should live in the `api` package. The `api` package should not have to know about the `crypto` package.

This already is a great improvement on separation though. We can improve further later on if that's easier.

Reply:
We could move it to `internals/secrethub` for now.

Reply:
This is pure laziness. Calling `.ToCiphertextRSAAES()` is just too easy.